### ISSUE TITLE
IWYU can't decide whether to include or not

### ISSUE THREAD

**Issue by martinlicht** [include-what-you-use/include-what-you-use#823]
IWYU shows inconsistent behavior of IWYU in one of my projects. I can reproduce it with the following code snippet.

```
//test.hpp
#include <algorithm>
#include <functional>

template<typename T>
void apply_sort_and_then_unique( T& t ) {
    std::sort( t.begin(), t.end() );
    auto last = std::unique( t.begin(), t.end() );
    t.erase( last, t.end() );
}
```
Running 'iwyu --std=c++17 test.hpp' gives 

```
$ iwyu --std=c++17 test.hpp 

test.hpp should add these lines:

test.hpp should remove these lines:
- #include <algorithm>  // lines 1-1
- #include <functional>  // lines 2-2

The full include-list for test.hpp:
---
```
Removing them and running IWYU again gives a compilation error. Removing the inclusion of `functional` still leaves the suggestion to remove `algorithm` as well. But leaving `functional` included and removing `algorithm` gives:
```
$ iwyu --std=c++17 test.hpp 

test.hpp should add these lines:
#include <algorithm>  // for sort, unique

test.hpp should remove these lines:
- #include <functional>  // lines 2-2

The full include-list for test.hpp:
#include <algorithm>  // for sort, unique
---
```

For your information:

```
$ iwyu --version
include-what-you-use 0.12 based on clang version 9.0.1-10
$ clang++ --version
clang version 10.0.0-4ubuntu1 
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
IWYU is as installed from Ubuntu repositories by the time of this posting.

**Comment by vintagepc** [comment]
I've noticed this with other C++ STL headers like <fsteam> and <sstream> too.

**Comment by kimgr** [comment]
Could you run this with higher verbosity level and see what the decision chain looks like?

**Comment by ts826848** [comment]
I think I've noticed something similar. I think it might have to do with the fact that it's dealing with templates that haven't been instantiated?

For example, given this simple template:

```cpp
// test.cpp
#include <type_traits>
template<typename T>
struct wrapper {
	constexpr static bool b = std::is_unsigned_v<T>;
};
```

IWYU says that `type_traits` should be removed:

```
bc@r2 /tmp % include-what-you-use -std=c++17 test.cpp

test.cpp should add these lines:

test.cpp should remove these lines:
- #include <type_traits>  // lines 2-2

The full include-list for test.cpp:
---
```

Making the suggested change results in a compilation error, though:

```
bc@r2 /tmp % include-what-you-use -std=c++17 test.cpp
test.cpp:4:28: error: use of undeclared identifier 'std'
        constexpr static bool b = std::is_unsigned_v<T>;
                                  ^
test.cpp:4:47: error: 'T' does not refer to a value
        constexpr static bool b = std::is_unsigned_v<T>;
                                                     ^
test.cpp:2:19: note: declared here
template<typename T>
                  ^
test.cpp:4:49: error: expected expression
        constexpr static bool b = std::is_unsigned_v<T>;
```

However, if the template is explicitly instantiated, IWYU no longer complains:

```cpp
// test.cpp
#include <type_traits>
template<typename T>
struct wrapper {
	constexpr static bool b = std::is_unsigned_v<T>;
};
template class wrapper<int>;
```

```
bc@r2 /tmp % include-what-you-use -std=c++17 test.cpp

(test.cpp has correct #includes/fwd-decls)
```

Verbose invocation of the explicit instantiation-less version:

```
bc@r2 /tmp % include-what-you-use -v -std=c++17 test.cpp
clang version 11.1.0
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /usr/local/Cellar/include-what-you-use/0.15_2/libexec/bin
clang invocation:
 "/usr/local/Cellar/include-what-you-use/0.15_2/libexec/bin/include-what-you-use" "-cc1" "-triple" "x86_64-apple-macosx10.15.0" "-Wundef-prefix=TARGET_OS_" "-Werror=undef-prefix" "-Wdeprecated-objc-isa-usage" "-Werror=deprecated-objc-isa-usage" "-fsyntax-only" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "test.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mframe-pointer=all" "-fno-rounding-math" "-munwind-tables" "-fcompatibility-qualified-id-block-type-checking" "-target-cpu" "penryn" "-debugger-tuning=lldb" "-target-linker-version" "609.8" "-v" "-resource-dir" "/usr/local/Cellar/include-what-you-use/0.15_2/libexec/lib/clang/11.1.0" "-isysroot" "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk" "-stdlib=libc++" "-internal-isystem" "/usr/local/Cellar/include-what-you-use/0.15_2/libexec/bin/../include/c++/v1" "-internal-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/c++/v1" "-internal-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/local/include" "-internal-isystem" "/usr/local/Cellar/include-what-you-use/0.15_2/libexec/lib/clang/11.1.0/include" "-internal-externc-isystem" "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include" "-std=c++17" "-fdeprecated-macro" "-fdebug-compilation-dir" "/tmp" "-ferror-limit" "19" "-stack-protector" "1" "-fblocks" "-fencode-extended-block-signature" "-fregister-global-dtors-with-atexit" "-fgnuc-version=4.2.1" "-fcxx-exceptions" "-fexceptions" "-fmax-type-align=16" "-x" "c++" "test.cpp"

clang -cc1 version 11.1.0 based upon LLVM 11.1.0 default target x86_64-apple-darwin19.6.0
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/c++/v1"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/local/include"
ignoring nonexistent directory "/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/Cellar/include-what-you-use/0.15_2/libexec/bin/../include/c++/v1
 /usr/local/Cellar/include-what-you-use/0.15_2/libexec/lib/clang/11.1.0/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include
 /Library/Developer/CommandLineTools/usr/include/c++/v1
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1
 /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks (framework directory)
End of search list.

test.cpp should add these lines:

test.cpp should remove these lines:
- #include <type_traits>  // lines 2-2

The full include-list for test.cpp:
---
```

The verbose invocation for the code with explicit instantiation results in precisely the same output, except no complaints about headers that should be added/removed.

**Comment by kimgr** [comment]
I think the key is:
```
test.cpp:5:28: (4) [ UnresolvedLookupExpr ] 0x7fc5348c1a30 UnresolvedLookupExpr 0x7fc5348c1a30 '<dependent type>' lvalue (no ADL) = 'is_unsigned_v' 0x7fc534812958
``` 

See also #636, which seems to have the same root cause.

**Comment by kimgr** [comment]
That said, I'm not sure the original report in this issue has the same root cause.

**Comment by ts826848** [comment]
For what it's worth, here's the AST and following lines from the verbose invocation for the original test case of this issue (comment is too long if I include everything):

```
bc@r2 /tmp % include-what-you-use -Xiwyu -v7 -std=c++17 test.hpp
Setting verbose-level to 7

<snip>

test.hpp:6:6: (1) [ FunctionTemplateDecl ] 0x7fdadb8c3e70 template <typename T> void apply_sort_and_then_unique(T &t)
test.hpp:5:19: (2) [ TemplateTypeParmDecl ] 0x7fdadb8c3c18 typename T
test.hpp:5:19: (3) [ TemplateTypeParmType ] 0x7fdadb8c3c70 T
test.hpp:6:6: (2) [ FunctionDecl ] 0x7fdadb8c3dd0 void apply_sort_and_then_unique(T &t)
test.hpp:6:1: (3) [ FunctionProtoTypeLoc ] void (T &)
test.hpp:6:1: (3) [ FunctionProtoType ] 0x7fdadb8c3d70 void (T &)
test.hpp:6:1: (4) [ BuiltinTypeLoc ] void
test.hpp:6:1: (4) [ BuiltinType ] 0x7fdad7829460 void
test.hpp:6:37: (4) [ ParmVarDecl ] 0x7fdadb8c3d00 T &t
test.hpp:6:34: (5) [ LValueReferenceTypeLoc ] T &
test.hpp:6:34: (5) [ LValueReferenceType ] 0x7fdadb8c3cc0 T &
test.hpp:6:34: (6) [ TemplateTypeParmTypeLoc ] T
test.hpp:6:34: (6) [ TemplateTypeParmType ] 0x7fdadb8c3c70 T
test.hpp:6:41: (3) [ CompoundStmt ] 0x7fdadb8c44b8 CompoundStmt 0x7fdadb8c44b8
|-CallExpr 0x7fdadb8c40e0 '<dependent type>'
| |-UnresolvedLookupExpr 0x7fdadb8c3f68 '<overloaded function type>' lvalue (no ADL) = 'sort' 0x7fdadb85fec0 0x7fdadb860780 0x7fdadb860fa0 0x7fdadb8629b0 0x7fdadb8632b0
| |-CallExpr 0x7fdadb8c4038 '<dependent type>'
| | `-CXXDependentScopeMemberExpr 0x7fdadb8c3ff0 '<dependent type>' lvalue .begin
| |   `-DeclRefExpr 0x7fdadb8c3fd0 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
| `-CallExpr 0x7fdadb8c40c0 '<dependent type>'
|   `-CXXDependentScopeMemberExpr 0x7fdadb8c4078 '<dependent type>' lvalue .end
|     `-DeclRefExpr 0x7fdadb8c4058 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
|-DeclStmt 0x7fdadb8c4360
| `-VarDecl 0x7fdadb8c4148  referenced last 'auto' cinit
|   `-CallExpr 0x7fdadb8c4330 '<dependent type>'
|     |-UnresolvedLookupExpr 0x7fdadb8c41d0 '<overloaded function type>' lvalue (no ADL) = 'unique' 0x7fdadb804c50 0x7fdadb8059f0
|     |-CallExpr 0x7fdadb8c4288 '<dependent type>'
|     | `-CXXDependentScopeMemberExpr 0x7fdadb8c4240 '<dependent type>' lvalue .begin
|     |   `-DeclRefExpr 0x7fdadb8c4220 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
|     `-CallExpr 0x7fdadb8c4310 '<dependent type>'
|       `-CXXDependentScopeMemberExpr 0x7fdadb8c42c8 '<dependent type>' lvalue .end
|         `-DeclRefExpr 0x7fdadb8c42a8 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
`-CallExpr 0x7fdadb8c4488 '<dependent type>'
  |-CXXDependentScopeMemberExpr 0x7fdadb8c4398 '<dependent type>' lvalue .erase
  | `-DeclRefExpr 0x7fdadb8c4378 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
  |-DeclRefExpr 0x7fdadb8c43e0 'auto' lvalue Var 0x7fdadb8c4148 'last' 'auto'
  `-CallExpr 0x7fdadb8c4468 '<dependent type>'
    `-CXXDependentScopeMemberExpr 0x7fdadb8c4420 '<dependent type>' lvalue .end
      `-DeclRefExpr 0x7fdadb8c4400 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:7:5: (4) [ CallExpr ] 0x7fdadb8c40e0 CallExpr 0x7fdadb8c40e0 '<dependent type>'
|-UnresolvedLookupExpr 0x7fdadb8c3f68 '<overloaded function type>' lvalue (no ADL) = 'sort' 0x7fdadb85fec0 0x7fdadb860780 0x7fdadb860fa0 0x7fdadb8629b0 0x7fdadb8632b0
|-CallExpr 0x7fdadb8c4038 '<dependent type>'
| `-CXXDependentScopeMemberExpr 0x7fdadb8c3ff0 '<dependent type>' lvalue .begin
|   `-DeclRefExpr 0x7fdadb8c3fd0 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
`-CallExpr 0x7fdadb8c40c0 '<dependent type>'
  `-CXXDependentScopeMemberExpr 0x7fdadb8c4078 '<dependent type>' lvalue .end
    `-DeclRefExpr 0x7fdadb8c4058 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:7:5: (5) [ UnresolvedLookupExpr ] 0x7fdadb8c3f68 UnresolvedLookupExpr 0x7fdadb8c3f68 '<overloaded function type>' lvalue (no ADL) = 'sort' 0x7fdadb85fec0 0x7fdadb860780 0x7fdadb860fa0 0x7fdadb8629b0 0x7fdadb8632b0

Adding to processed_overload_locs: test.hpp:7:5
Marked full-info use of decl 0x7fdadb85fe20 std::sort (from /usr/local/Cellar/include-what-you-use/0.15_2/libexec/include/c++/v1/algorithm:4122:1) at test.hpp:7:5
test.hpp:7:5: (6) [ NestedNameSpecifier ] 0x7fdad87d7aa8 std::
test.hpp:7:16: (5) [ CallExpr ] 0x7fdadb8c4038 CallExpr 0x7fdadb8c4038 '<dependent type>'
`-CXXDependentScopeMemberExpr 0x7fdadb8c3ff0 '<dependent type>' lvalue .begin
  `-DeclRefExpr 0x7fdadb8c3fd0 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:7:17: (6) [ CXXDependentScopeMemberExpr ] 0x7fdadb8c3ff0 CXXDependentScopeMemberExpr 0x7fdadb8c3ff0 '<dependent type>' lvalue .begin
`-DeclRefExpr 0x7fdadb8c3fd0 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:7:16: (7) [ DeclRefExpr ] 0x7fdadb8c3fd0 DeclRefExpr 0x7fdadb8c3fd0 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

Marked full-info use of decl 0x7fdadb8c3d00 t (from test.hpp:6:37) at test.hpp:7:16
test.hpp:7:27: (5) [ CallExpr ] 0x7fdadb8c40c0 CallExpr 0x7fdadb8c40c0 '<dependent type>'
`-CXXDependentScopeMemberExpr 0x7fdadb8c4078 '<dependent type>' lvalue .end
  `-DeclRefExpr 0x7fdadb8c4058 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:7:28: (6) [ CXXDependentScopeMemberExpr ] 0x7fdadb8c4078 CXXDependentScopeMemberExpr 0x7fdadb8c4078 '<dependent type>' lvalue .end
`-DeclRefExpr 0x7fdadb8c4058 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:7:27: (7) [ DeclRefExpr ] 0x7fdadb8c4058 DeclRefExpr 0x7fdadb8c4058 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

Marked full-info use of decl 0x7fdadb8c3d00 t (from test.hpp:6:37) at test.hpp:7:27
test.hpp:8:5: (4) [ DeclStmt ] 0x7fdadb8c4360 DeclStmt 0x7fdadb8c4360
`-VarDecl 0x7fdadb8c4148  referenced last 'auto' cinit
  `-CallExpr 0x7fdadb8c4330 '<dependent type>'
    |-UnresolvedLookupExpr 0x7fdadb8c41d0 '<overloaded function type>' lvalue (no ADL) = 'unique' 0x7fdadb804c50 0x7fdadb8059f0
    |-CallExpr 0x7fdadb8c4288 '<dependent type>'
    | `-CXXDependentScopeMemberExpr 0x7fdadb8c4240 '<dependent type>' lvalue .begin
    |   `-DeclRefExpr 0x7fdadb8c4220 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
    `-CallExpr 0x7fdadb8c4310 '<dependent type>'
      `-CXXDependentScopeMemberExpr 0x7fdadb8c42c8 '<dependent type>' lvalue .end
        `-DeclRefExpr 0x7fdadb8c42a8 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:8:10: (5) [ VarDecl ] 0x7fdadb8c4148 auto last
test.hpp:8:5: (6) [ AutoTypeLoc ] auto
test.hpp:8:5: (6) [ AutoType ] 0x7fdada818550 auto
test.hpp:8:17: (6) [ CallExpr ] 0x7fdadb8c4330 CallExpr 0x7fdadb8c4330 '<dependent type>'
|-UnresolvedLookupExpr 0x7fdadb8c41d0 '<overloaded function type>' lvalue (no ADL) = 'unique' 0x7fdadb804c50 0x7fdadb8059f0
|-CallExpr 0x7fdadb8c4288 '<dependent type>'
| `-CXXDependentScopeMemberExpr 0x7fdadb8c4240 '<dependent type>' lvalue .begin
|   `-DeclRefExpr 0x7fdadb8c4220 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
`-CallExpr 0x7fdadb8c4310 '<dependent type>'
  `-CXXDependentScopeMemberExpr 0x7fdadb8c42c8 '<dependent type>' lvalue .end
    `-DeclRefExpr 0x7fdadb8c42a8 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:8:17: (7) [ UnresolvedLookupExpr ] 0x7fdadb8c41d0 UnresolvedLookupExpr 0x7fdadb8c41d0 '<overloaded function type>' lvalue (no ADL) = 'unique' 0x7fdadb804c50 0x7fdadb8059f0

Adding to processed_overload_locs: test.hpp:8:17
Marked full-info use of decl 0x7fdadb804bb0 std::unique (from /usr/local/Cellar/include-what-you-use/0.15_2/libexec/include/c++/v1/algorithm:2176:1) at test.hpp:8:17
test.hpp:8:17: (8) [ NestedNameSpecifier ] 0x7fdad87d7aa8 std::
test.hpp:8:30: (7) [ CallExpr ] 0x7fdadb8c4288 CallExpr 0x7fdadb8c4288 '<dependent type>'
`-CXXDependentScopeMemberExpr 0x7fdadb8c4240 '<dependent type>' lvalue .begin
  `-DeclRefExpr 0x7fdadb8c4220 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:8:31: (8) [ CXXDependentScopeMemberExpr ] 0x7fdadb8c4240 CXXDependentScopeMemberExpr 0x7fdadb8c4240 '<dependent type>' lvalue .begin
`-DeclRefExpr 0x7fdadb8c4220 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:8:30: (9) [ DeclRefExpr ] 0x7fdadb8c4220 DeclRefExpr 0x7fdadb8c4220 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

Marked full-info use of decl 0x7fdadb8c3d00 t (from test.hpp:6:37) at test.hpp:8:30
test.hpp:8:41: (7) [ CallExpr ] 0x7fdadb8c4310 CallExpr 0x7fdadb8c4310 '<dependent type>'
`-CXXDependentScopeMemberExpr 0x7fdadb8c42c8 '<dependent type>' lvalue .end
  `-DeclRefExpr 0x7fdadb8c42a8 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:8:42: (8) [ CXXDependentScopeMemberExpr ] 0x7fdadb8c42c8 CXXDependentScopeMemberExpr 0x7fdadb8c42c8 '<dependent type>' lvalue .end
`-DeclRefExpr 0x7fdadb8c42a8 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:8:41: (9) [ DeclRefExpr ] 0x7fdadb8c42a8 DeclRefExpr 0x7fdadb8c42a8 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

Marked full-info use of decl 0x7fdadb8c3d00 t (from test.hpp:6:37) at test.hpp:8:41
test.hpp:9:5: (4) [ CallExpr ] 0x7fdadb8c4488 CallExpr 0x7fdadb8c4488 '<dependent type>'
|-CXXDependentScopeMemberExpr 0x7fdadb8c4398 '<dependent type>' lvalue .erase
| `-DeclRefExpr 0x7fdadb8c4378 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'
|-DeclRefExpr 0x7fdadb8c43e0 'auto' lvalue Var 0x7fdadb8c4148 'last' 'auto'
`-CallExpr 0x7fdadb8c4468 '<dependent type>'
  `-CXXDependentScopeMemberExpr 0x7fdadb8c4420 '<dependent type>' lvalue .end
    `-DeclRefExpr 0x7fdadb8c4400 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:9:6: (5) [ CXXDependentScopeMemberExpr ] 0x7fdadb8c4398 CXXDependentScopeMemberExpr 0x7fdadb8c4398 '<dependent type>' lvalue .erase
`-DeclRefExpr 0x7fdadb8c4378 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:9:5: (6) [ DeclRefExpr ] 0x7fdadb8c4378 DeclRefExpr 0x7fdadb8c4378 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

Marked full-info use of decl 0x7fdadb8c3d00 t (from test.hpp:6:37) at test.hpp:9:5
test.hpp:9:14: (5) [ DeclRefExpr ] 0x7fdadb8c43e0 DeclRefExpr 0x7fdadb8c43e0 'auto' lvalue Var 0x7fdadb8c4148 'last' 'auto'

Marked full-info use of decl 0x7fdadb8c4148 last (from test.hpp:8:10) at test.hpp:9:14
test.hpp:9:20: (5) [ CallExpr ] 0x7fdadb8c4468 CallExpr 0x7fdadb8c4468 '<dependent type>'
`-CXXDependentScopeMemberExpr 0x7fdadb8c4420 '<dependent type>' lvalue .end
  `-DeclRefExpr 0x7fdadb8c4400 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:9:21: (6) [ CXXDependentScopeMemberExpr ] 0x7fdadb8c4420 CXXDependentScopeMemberExpr 0x7fdadb8c4420 '<dependent type>' lvalue .end
`-DeclRefExpr 0x7fdadb8c4400 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

test.hpp:9:20: (7) [ DeclRefExpr ] 0x7fdadb8c4400 DeclRefExpr 0x7fdadb8c4400 'T' lvalue ParmVar 0x7fdadb8c3d00 't' 'T &'

Marked full-info use of decl 0x7fdadb8c3d00 t (from test.hpp:6:37) at test.hpp:9:20
--- Calculating IWYU violations for test.hpp ---
Ignoring use of t (test.hpp:7:16): definition is present: test.hpp:6:37
Ignoring use of t (test.hpp:7:27): definition is present: test.hpp:6:37
Ignoring use of t (test.hpp:8:30): definition is present: test.hpp:6:37
Ignoring use of t (test.hpp:8:41): definition is present: test.hpp:6:37
Ignoring use of t (test.hpp:9:5): definition is present: test.hpp:6:37
Ignoring use of last (test.hpp:9:14): definition is present: test.hpp:8:10
Ignoring use of t (test.hpp:9:20): definition is present: test.hpp:6:37
Mapped /usr/local/Cellar/include-what-you-use/0.15_2/libexec/include/c++/v1/algorithm to <algorithm> for std::sort (only candidate)
Mapped /usr/local/Cellar/include-what-you-use/0.15_2/libexec/include/c++/v1/algorithm to <algorithm> for std::unique (only candidate)
Ignoring full use of std::sort (test.hpp:7:5): #including dfn from <algorithm>
Ignoring full use of std::unique (test.hpp:8:17): #including dfn from <algorithm>

test.hpp should add these lines:

test.hpp should remove these lines:
- #include <functional>  // lines 3-3

The full include-list for test.hpp:
#include <algorithm>  // for sort, unique
---
```

**Comment by ptomato** [comment]
Can confirm this with iwyu 0.16:

**bug.cpp:**
```c++
#include <algorithm>
#include <vector>
template <typename T>
void a(std::vector<T> v) {
    std::find(v.begin(), v.end(), T{});
}
```

```
$ include-what-you-use -std=c++17 bug.cpp

bug.cpp should add these lines:

bug.cpp should remove these lines:
- #include <algorithm>  // lines 1-1

The full include-list for bug.cpp:
#include <vector>  // for vector
---
```

Unlike in the earlier example, explicitly instantiating `template void a<int>(std::vector<int>);` does not fix the problem.

Also worth noting that `-std=c++17` is required to reproduce the problem. It doesn't show up with `-std=c++14`.

**Comment by kimgr** [comment]
This is fixed as of #1511 (cad1a68c768c2b27ec7bba37e3e9f39456b7bc3e on master), except for the variant reported by @ts826848 -- I extracted a separate issue out for that one: #1516.
